src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
+{
+    public class ResolveGenericsSyntax :
+        SyntaxTreeTransformation<ResolveGenericsScope>
+    {
+        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
+        {
+            if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
+
+            // Get list of generics
+            List<QsQualifiedName> generics = new List<QsQualifiedName>();
+            var Callables = namespaces.GlobalCallableResolutions();
+            foreach (var callable in Callables)
+            {
+                if (callable.Value.Signature.TypeParameters.Any())
+                {
+                    generics.Add(callable.Key);
+                }
+            }
+
+            var filter = new ResolveGenericsSyntax(new ResolveGenericsScope());
+            return namespaces.Select(ns => filter.Transform(ns));
+        }
+
+        public static IEnumerable<QsNamespace> Apply(params QsNamespace[] namespaces)
+        {
+            return Apply(namespaces);
+        }
+
+        public ResolveGenericsSyntax(ResolveGenericsScope scope) : base(scope) { }
+    }
+
+    public class ResolveGenericsScope :
+            ScopeTransformation<ResolveGenericsExpression>
+    {
+        public ResolveGenericsScope() : base(new ResolveGenericsExpression()) { }
+    }
+
+    public class ResolveGenericsExpression :
+        ExpressionTransformation<ResolvedGenericsExpressionKind>
+    {
+        public ResolveGenericsExpression() : base(ex => new ResolvedGenericsExpressionKind(ex as ResolveGenericsExpression)) { }
+
+        public override TypedExpression Transform(TypedExpression ex)
+        {
+            ImmutableDictionary<QsTypeParameter, ResolvedType> types = ex.TypeParameterResolutions;
+
+            if (types.Any() && ex.Expression is QsExpressionKind<TypedExpression, Identifier, ResolvedType>.CallLikeExpression call)
+            {
+                string prefix = CreateTypePrefix(types);
+
+                // For now, only resolve identifiers of global callables
+                if (call.Item1.Expression is QsExpressionKind<TypedExpression, Identifier, ResolvedType>.Identifier id &&
+                    id.Item1 is Identifier.GlobalCallable globalCallable)
+                {
+                    // Rebuild the method with the updated identifier
+                    ex = new TypedExpression(
```

Comment:
Is there a framework I could leverage to create new objects from existing objects with only one field changed? This block of code is rather unpleasant.

Reply:
Yes, there are some helper functions around. Let's get together and go over this together.

Reply:
This will need to be addressed later once there is a data structure for keeping tabs on information regarding generics.